crates/sparrow-merge/src/merge_pipeline.rs

```diff
@@ -0,0 +1,340 @@
+use arrow_schema::DataType;
+use error_stack::{IntoReport, ResultExt};
+use parking_lot::Mutex;
+use sparrow_batch::Batch;
+use sparrow_physical::StepId;
+use sparrow_scheduler::{
+    InputHandles, Partition, Partitioned, Pipeline, PipelineError, Scheduler, TaskRef,
+};
+use std::sync::atomic::{AtomicBool, Ordering};
+use tokio::sync::mpsc::error::TryRecvError;
+
+use crate::merge::HeterogeneousMerge;
+
+/// Runs a merge operation.
+pub struct MergePipeline {
+    /// The state for each partition.
+    partitions: Partitioned<MergePartition>,
+    /// Input handles for the consuming (receiving) computations.
+    consumers: InputHandles,
+    /// The left input.
+    left: Input,
+    /// The right input.
+    right: Input,
+    /// The result type of this merge.
+    result_type: DataType,
+}
+
+impl std::fmt::Debug for MergePipeline {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MergePipeline").finish()
+    }
+}
+
+struct Input {
+    /// The input id for this input
+    #[allow(unused)]
+    id: StepId,
+    /// The data type for this input
+    datatype: DataType,
+    /// Whether this side is closed
+    is_closed: Mutex<bool>,
```

Comment:
I think `AtomicBool` is often preferred to `Mutex<bool>` (https://users.rust-lang.org/t/is-it-okay-to-use-mutex-bool/90742) but this is probably OK. The main risk is that we need to be careful of lock order if we ever lock both this and some other mutex, to always do so in the same order.